Adding the tutorial 'Customize Delegate to Run Your Favorite Third Party Tools' (#520)
harness/developer-hub: tutorials/platform/customize-delegate.md

> # Customize the Delegate to Run Your Favorite Third-Party Tools

Reviewer: ```suggestion
# Customize the delegate to run your favorite third-party tools
```

In this tutorial, we'll show you how to install Harness Delegate, and how to customize the delegate to run your favorite tools. 

## Prerequisites

- A free [Harness cloud](https://app.harness.io/auth/#/signup/utm_source=website&utm_medium=harness-developer-hub&utm_campaign=cd-plg&utm_content=get-started) account 
- A deployment target such as Kubernetes cluster. You can use Minikube or Kind to create a cluster. In this tutorial, we'll use a cluster that was created on Google Cloud.
  
### Tutorial

Now that you've got a Harness cloud account, you can create a project.

![about the project](./static/customizing-delegate/about_project.png)

You can start by inviting collaborators to work with you, or you can click **Save and continue**. 

![invite collaborators](./static/customizing-delegate/collaborators.png)

> Next, you need to pick a Harness module for your project. Click the **Continuous Delivery** card.

Reviewer: ```suggestion
Next, pick a Harness module for your project. Click the **Continuous Delivery** card.
```

![delegate running](./static/customizing-delegate/delegate_up_running.png)

Let’s create a pipeline and add a custom stage

![adding custom stage](./static/customizing-delegate/custom_stage.png)

Add a step and select **Shell Script** from the options.

![adding the stage](./static/customizing-delegate/add_step.png)

![adding step library](./static/customizing-delegate/step_library.png)

Let’s add the command to check the Git version installed on the delegate.

Add the command `git --version` in the Script tab. 

![checking git version script](./static/customizing-delegate/git_version_shell.png)

Make sure the Execution Target is set to **On Delegate**. 

In the **Advanced** tab, select the delegate that's running and where Git is installed.

![advanced tab](./static/customizing-delegate/shell_script_advanced.png)

Now, apply your changes, save everything, and run the pipeline.

![run pipeline step](./static/customizing-delegate/run_pipeline_step.png)

You should see the Git command execution. 

![git command execution](./static/customizing-delegate/git_command_execution.png)

Now, you can use any Git command in your project. 

### Install AWS CLI

> You can add the AWS CLI tooling on your Delegate just the way we added Git.

Reviewer: ```suggestion
You can add AWS CLI tooling to your delegate the same way we added Git.
```

Here is the AWS CLI script to add into our `INIT_SCRIPT` section:

```
curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip"
unzip awscliv2.zip
./aws/install
```

See the screenshot below for more clarity.

![aws cli step](./static/customizing-delegate/aws_cli_script.png)

> Let’s add the shell script as a step just like how we added Git.

Reviewer: ```suggestion
Let’s add the shell script as a step the same way we added Git.
```

In the script, add the command `aws –version`
![aws cli version check](./static/customizing-delegate/aws_version.png)

Also, specify the delegate in the advanced settings.
![specify delegate](./static/customizing-delegate/delegate_advanced_settings.png)

Save everything and run the pipeline. 

![running the pipeline](./static/customizing-delegate/aws_run_pipeline.png)
You can see the AWS CLI version.

### Install kubectl

> Add the following script into your script and apply the yaml just like how we did previously.

Reviewer: ```suggestion
Add the following script into your script and apply the YAML the same way we did before.
```

```
curl -L0 https://dl.k8s.io/release/v1.24.3/bin/linux/amd64/kubectl -o kubectl  
chmod +x ./kubectl  
mv kubectl /opt/harness-delegate/custom-client-tools/kubectl
```

See the screenshot below for more clarity.

![kubectl script image](./static/customizing-delegate/kubectl_init_script.png)

Save the YAML and apply the `kubectl` command for the delegate.

> Add the shell script step as shown below with the `kubectl version` command.

Reviewer: ```suggestion
Add the shell script step as shown below. Use the `kubectl version` command.
```

![kubectl version adding](./static/customizing-delegate/kubectl_version_command.png)

> Make sure you specify the delegate on the **Advanced** tab of the above step.

Reviewer: ```suggestion
Make sure you specify the delegate on the **Advanced** tab of the step.
```

Save everything and run the pipeline.

![run the pipeline](./static/customizing-delegate/kubectl_run_pipeline.png)

The `kubectl` version information is displayed. 

### Install Terraform

Add this script to the `IN_IT` section:

```
curl -O -L  https://releases.hashicorp.com/terraform/0.12.25/terraform_0.12.25_linux_amd64.zip  
unzip terraform_0.12.25_linux_amd64.zip  
mv ./terraform /usr/bin/  
```

Save the YAML and apply the `kubectl` command.

> Add the shell script step as shown below, with the `terraform --version` command.

Reviewer: ```suggestion
Add the shell script step as follows. Use the `terraform --version` command.
```

![checking Helm version](./static/customizing-delegate/helm_version_execution.png)

Let’s verify that we can deploy our charts.

In the Shell Script, replace the existing command with the following:

```
helm create my-new-chart
helm install my-new-chart ./my-new-chart
helm ls
```
Apply changes, save the configuration, and run the pipeline. Your chart should be deployed.

![deploying helm chart](./static/customizing-delegate/helm_chart_deployed.png)

### Install all at once

Let’s combine all the tools and add them all to the `IN_IT` script. 

![Installing all at once](./static/customizing-delegate/all_at_once.png)

Save the changes and use `kubectl` to apply the YAML to the delegate.

You can create everything from one step or create separate steps for each tool. In this tutorial, we created separate steps for each tool.

![separate step](./static/customizing-delegate/seperate_for_each.png)

Save the configuration and run the pipeline. On success, you should see output that includes the different tool version numbers. 

![running the pipeline successfully](./static/customizing-delegate/together_execution.png)

> This way, you can install and play with any of your favorite tools on Harness Delegate. This customization gives a lot of flexibility to developers working with these tools.  

Reviewer: ```suggestion
You can use these steps to install and play with any of your favorite tools on Harness Delegate. This customization gives you a lot of flexibility.  
```
